src/pcp.rs

```diff
@@ -299,114 +304,152 @@ where
+///
+/// TODO(cjpatton) Consider extending this API with a `Param` associated type and add a trait bound
+/// for trying to construct an instance of from parameters and the number of times the gadget will
+/// be called.
+pub trait Gadget<F: FieldElement> {
     /// Evaluates the gadget on input `inp` and returns the output.
     fn call(&mut self, inp: &[F]) -> Result<F, PcpError>;
 
-    /// Returns the circuit's arity, i.e., the expected length of the input to `call`.
-    fn call_in_len(&self) -> usize;
-}
-
-/// The sub-circuit associated with some validity circuit. A gadget is called either on a sequence
-/// of finite field elements or a sequence of polynomials over a finite field.
-pub trait Gadget<F: FieldElement>: GadgetCallOnly<F> {
     /// Evaluate the gadget on input of a sequence of polynomials. The output is written to `outp`.
-    fn call_poly<V: AsRef<[F]>>(&mut self, outp: &mut [F], inp: &[V]) -> Result<(), PcpError>;
+    fn call_poly(&mut self, outp: &mut [F], inp: &Vec<Vec<F>>) -> Result<(), PcpError>;
+
+    /// Returns the arity of the gadget. This is the length of `inp` passed to `call` or
+    /// `call_poly`.
+    fn arity(&self) -> usize;
 
-    /// The maximum degree of the polynomial output by `call_poly` as a function of the maximum
-    /// degree of each input polynomial.
-    fn call_poly_out_len(&self, in_len: usize) -> usize;
+    /// Returns the circuit's arithmetic degree. This determines the minimum length the `outp`
+    /// buffer passed to `call_poly`.
+    fn deg(&self) -> usize;
+
+    /// This call is used to downcast a `Box<dyn Gadget<F>>` to a concrete type.
+    fn as_any(&mut self) -> &mut dyn Any;
 }
 
 /// Generate a proof of an input's validity.
-pub fn prove<F, G, V>(x: &V, joint_rand: &[F]) -> Result<Proof<F>, PcpError>
+pub fn prove<F, V>(x: &V, joint_rand: &[F]) -> Result<Proof<F>, PcpError>
 where
     F: FieldElement,
-    G: Gadget<F>,
-    V: Value<F, G>,
+    V: Value<F>,
 {
     let g_calls = x.valid_gadget_calls();
-    let m = (g_calls + 1).next_power_of_two();
-    let mut g = x.gadget(m);
-    let p = g.call_poly_out_len(m);
-    let l = g.call_in_len();
-    let mut data = vec![F::zero(); l + p];
-
-    // Run the validity circuit with a "shim" gadget that records the value of each input wire of
-    // each gadget evaluation.
-    let mut shim = ProveShimGadget::new(&mut g, g_calls)?;
+
+    let mut shim = x
+        .gadget()
+        .into_iter()
+        .enumerate()
+        .map(|(idx, g)| ProveShimGadget::new(g, g_calls[idx]))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Create a buffer for storing the proof. The buffer is longer than the proof itself; the extra
+    // length is to accommodate the computation of each of the proof polynomials.
+    let data_len = (0..shim.len())
+        .map(|idx| shim[idx].arity() + shim[idx].deg() * (1 + g_calls[idx]).next_power_of_two())
+        .reduce(|a, b| a + b)
+        .unwrap();
```

Comment:
```suggestion
        .sum();
```
`Iterator::sum()` doesn't return an `Option` so you can remove the unsightly `unwrap()`

Reply:
Nice!